jbpm-workitems/jbpm-workitems-email/src/test/java/org/jbpm/process/workitem/email/TemplateManagerTest.java

```diff
@@ -0,0 +1,165 @@
+ * limitations under the License.
+ */
+
+package org.jbpm.process.workitem.email;
+
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
+import static org.junit.Assert.assertEquals;
+
+import java.io.File;
+import java.nio.file.Files;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class TemplateManagerTest {
+    
+    private TemplateManager templateManager;
+    private File templateDir = new File("src/test/resources/templates");
+    
+    private String originalTemplate = "<html><body>Hello ${Name}</body></html>";
+    
+    @Before
+    public void setup() {
+        System.setProperty("org.jbpm.email.templates.watcher.enabled", "true");  
+        System.setProperty("org.jbpm.email.templates.dir", templateDir.getAbsolutePath());
+    }
+    
+    @After
+    public void cleanup() {
+        if (templateManager != null) {
+            templateManager.close();
+        }
+        
+        File[] templatesToRemove = templateDir.listFiles((file, name) -> { return !name.equals("basic-email.html");});
+        for (File template : templatesToRemove) {
+            template.delete();
+        }
+        
+        System.clearProperty("org.jbpm.email.templates.watcher.enabled");
+        System.clearProperty("org.jbpm.email.templates.dir");
+    }
+    
+    private void configureTemplateManager(CountDownLatch loadTemplate, CountDownLatch removeTemplate) {
+        templateManager = new TemplateManager(){
+
+            @Override
+            protected void loadTemplate(File templateFile) {
+                super.loadTemplate(templateFile);
+                if (loadTemplate != null) {
+                    loadTemplate.countDown();
+                }
+            }
+
+            @Override
+            protected void removeTemplate(File templateFile) {
+                super.removeTemplate(templateFile);
+                if (removeTemplate != null) {
+                    removeTemplate.countDown();
+                }
+            }
+            
+        };
+    }
+    
+    @Test(timeout=30000)
+    public void testLoadNewTemplateFile() throws Exception {
+        CountDownLatch loadTemplate = new CountDownLatch(2);// first is the initial load        
+        configureTemplateManager(loadTemplate, null);
+        
+        String expected = "<html><body>Hello John</body></html>";
+        
+        Map<String, Object> parameters = new HashMap<>();
+        parameters.put("Name", "John");
+        
+        String rendered = templateManager.render("basic-email", parameters);
+        assertEquals(expected, rendered);
+        
+        String advancedTemplate = "<html><body>Hello ${Name}, welcome!</body></html>";
+        String expectedAdvanced = "<html><body>Hello John, welcome!</body></html>";
+        
+        Thread.sleep(3000);
```

Comment:
I can see CountDownLatch is used for sync of template loading. What is this 3 seconds waiting required for?

Reply:
I added it to slightly delay the actual change but now removed it so let's see if that will work as expected